planet/data_filter.py

```diff
@@ -0,0 +1,118 @@
+# Copyright 2022 Planet Labs PBC.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not
+# use this file except in compliance with the License. You may obtain a copy of
+# the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+"""Functionality for preparing a data search filter"""
+from datetime import datetime
+import logging
+from typing import List, Union
+
+from planet import exceptions
+
+LOGGER = logging.getLogger(__name__)
+
+
+def and_filter(nested_filters: List[dict]) -> dict:
+    """Create an AndFilter
+
+    The AndFilter can be used to limit results to items with properties or
+    permissions which match all nested filters.
+
+    It is most commonly used as a top-level filter to ensure criteria across
+    all field and permission filters are met.
+
+    Parameters:
+        nested_filters: Filters to AND together
+    """
+    return {'type': 'AndFilter', 'config': nested_filters}
+
+
+def or_filter(nested_filters: List[dict]) -> dict:
+    """Create an OrFilter
+
+    The OrFilter can be used to match items with properties or permissions
+    which match at least one of the nested filters.
+
+    Parameters:
+        nested_filters: Filters to OR together
+    """
+    return {'type': 'OrFilter', 'config': nested_filters}
+
+
+def not_filter(nested_filter: dict):
+    """Create a NotFilter
+
+    The NotFilter can be used to match items with properties or permissions
+    which do not match the nested filter.
+
+    This filter only supports a single nested filter. Multiple NotFilters can
+    be nested within an AndFilter to filter across multiple fields or
+    permission values.
+
+    Parameters:
+        nested_filter: Filter to NOT
+    """
+    return {'type': 'NotFilter', 'config': nested_filter}
+
+
+def _field_filter(ftype: str, field_name: str, config: Union[dict,
+                                                             list]) -> dict:
+    return {'type': ftype, 'field_name': field_name, 'config': config}
+
+
+def date_range_filter(field_name: str,
```

Comment:
@sgillies I'm curious about your thoughts on this parameter structure for the optional conditionals. Data API ref: https://developers.planet.com/docs/apis/data/searches-filtering/#daterangefilter

Reply:
Do you mean my thoughts on assembling a config dict from named keyword args instead of taking a config dict as an arg? I feel like what you're doing shapes the input in a positive way and is user-friendly. If you wanted to future proof these filters against changes in operators, I guess you could add an optional `config: dict = None` arg and merge it with the other options?

Reply:
Thanks for your thoughts. That is a great suggestion. For now I feel fine with not worrying too much about future proofing, my primary concern is how user-friendly it is. If we're generally ok with it, I'll keep it like it is.